util/bloom/xor.go

```diff
@@ -0,0 +1,259 @@
+
+package bloom
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+
+	"github.com/algorand/xorfilter"
+)
+
+// XorBuilder is a local alias for xorfilter.Builder
+type XorBuilder = xorfilter.Builder
+
+// XorFilter is a faster more efficient alternative to a Bloom filter
+// An XorFilter object can be used as is or with optional adittional setup.
+type XorFilter struct {
+	xor     *xorfilter.Xor32
+	holding []uint64
+
+	b *XorBuilder
+}
+
+// NewXor returns an XorFilter with an internal map created with a size hint and an optional *XorBuilder (may be nil)
+// The Builder is not thread safe and should only be used by one thread at a time.
+func NewXor(hint int, builder *XorBuilder) *XorFilter {
+	return &XorFilter{
+		b: builder,
+	}
+}
+
+// Set adds the value to the filter.
+func (xf *XorFilter) Set(x []byte) {
+	k := binary.BigEndian.Uint64(x)
+	xf.holding = append(xf.holding, k)
+}
+
+// Test checks whether x is present in the filter.
+// May return (rare) erroneous true values, but false is precise.
+func (xf *XorFilter) Test(x []byte) bool {
+	k := binary.BigEndian.Uint64(x)
+	if xf.xor != nil {
+		return xf.xor.Contains(k)
+	}
+	return false
+}
+
+// MarshalBinary implements encoding.BinaryMarshaller interface
+func (xf *XorFilter) MarshalBinary() ([]byte, error) {
+	if len(xf.holding) != 0 {
+		var err error
+		if xf.b != nil {
+			xf.xor, err = xf.b.Populate32(xf.holding)
+		} else {
+			xf.xor, err = xorfilter.Populate32(xf.holding)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+	if xf.xor == nil || (len(xf.xor.Fingerprints) == 0) {
+		// TODO: some other encoding for empty set?
+		return nil, nil
+	}
+	out := make([]byte, binary.MaxVarintLen64+binary.MaxVarintLen32+binary.MaxVarintLen32+(len(xf.xor.Fingerprints)*4))
+	pos := 0
+	pos += binary.PutUvarint(out[pos:], xf.xor.Seed)
+	pos += binary.PutUvarint(out[pos:], uint64(xf.xor.BlockLength))
+	pos += binary.PutUvarint(out[pos:], uint64(len(xf.xor.Fingerprints)))
+	for _, v := range xf.xor.Fingerprints {
+		binary.LittleEndian.PutUint32(out[pos:], v)
+		pos += 4
+	}
+	out = out[:pos]
+	return out, nil
+}
+
+// ErrBadBinary is returned when UnmarshalBinary fails
+var ErrBadBinary = errors.New("bad XorFilter binary")
+
+// UnmarshalBinary implements encoding.BinaryUnmarshaller interface
+func (xf *XorFilter) UnmarshalBinary(data []byte) error {
+	pos := 0
+	var dp int
+	xor := new(xorfilter.Xor32)
+	xor.Seed, dp = binary.Uvarint(data[pos:])
+	if dp < 0 {
+		return ErrBadBinary
+	}
+	pos += dp
+	blockLength, dp := binary.Uvarint(data[pos:])
+	if dp < 0 {
+		return ErrBadBinary
+	}
+	xor.BlockLength = uint32(blockLength)
+	pos += dp
+	lenFingerprints, dp := binary.Uvarint(data[pos:])
+	if dp < 0 {
+		return ErrBadBinary
+	}
+	pos += dp
+	if lenFingerprints > 0 {
+		xor.Fingerprints = make([]uint32, lenFingerprints)
+		for i := 0; i < int(lenFingerprints); i++ {
+			xor.Fingerprints[i] = binary.LittleEndian.Uint32(data[pos:])
```

Comment:
not just. We need to add the following:
1. ( as you've suggested ) we need to ensure data has the *exact* number of bytes.
2. we need to limit the number of supported Fingerprints to avoid an attacker causing out of memory panic attack.
3. The method should return an error for a zero fingerprints : If we would allow to support zero finger prints, the transaction sync is going to test each transaction against the empty bloom filter. This is pointless and should be avoided by the caller. But in the current implementation, the caller doesn't know that it's an empty bloom filter.

Reply:
2. without making the `XorFilter` implementation more complex, we can just ensure we don't pass a very long byte array at the caller
3. It's not an attack vector, what will it really accomplish?